kafka-rest/src/main/java/io/confluent/kafkarest/controllers/PartitionManagerImpl.java

```diff
@@ -73,7 +79,23 @@ public CompletableFuture<Optional<Partition>> getPartition(
       String clusterId, String topicName, int partitionId) {
     return topicManager
         .getTopic(clusterId, topicName)
-        .thenApply(topic -> checkEntityExists(topic, "Topic %s cannot be found.", topic))
+        .handle(
+            (value, exception) -> {
+              if (exception != null) {
+                log.error("Caught exception while get topic from cluster", exception);
+                if (exception.getCause() instanceof UnknownTopicOrPartitionException) {
+                  String exceptionMessage =
+                      String.format(
+                          "This server does not host this topic-partition for topic %s", topicName);
+                  throw new UnknownTopicOrPartitionException(exceptionMessage, exception);
+                } else if (exception instanceof NotFoundException
+                    || exception.getCause() instanceof NotFoundException) {
+                  throw new NotFoundException(exception);
+                }
+                throw new RuntimeException(exception);
+              }
+              return checkEntityExists(value, "Topic %s cannot be found.", value);
+            })
         .thenApply(Topic::getPartitions)
```

Comment:
I think the problem was originally that the topic existed, but the specified partition did not exist any more - the customer was shrinking their cluster and it got stuck.

So, I think we need the same sort of logic in the ::getPartitions section too.